circuit/src/native.rs

```diff
@@ -0,0 +1,652 @@
+					ops_i.scores[j].1 = op_score * op_score_sum.invert().unwrap() * credits;
+				}
+			}
+		}
+
+		// There should be at least 2 valid peers(valid opinions) for calculation
+		let valid_peers = filtered_set.iter().filter(|(_, credits)| credits != &Fr::zero()).count();
+		assert!(valid_peers >= 2, "Insufficient peers for calculation!");
+
+		// By this point we should use filtered_set and filtered_opinions
+		let mut s = filtered_set.map(|item| item.1);
+		let default_op = Opinion::default();
+		for _ in 0..NUM_ITERATIONS {
+			let mut distributions = [[Fr::zero(); NUM_NEIGHBOURS]; NUM_NEIGHBOURS];
+			for i in 0..NUM_NEIGHBOURS {
+				let mut local_distr = [Fr::zero(); NUM_NEIGHBOURS];
+				let pk = filtered_set[i].0;
+				let ops_i = filtered_ops.get(&pk).unwrap_or_else(|| &default_op);
+
+				for j in 0..NUM_NEIGHBOURS {
+					let op = ops_i.scores[j].1 * s[i];
+					local_distr[j] = op;
+				}
+				distributions[i] = local_distr;
+			}
+
+			let mut new_s = [Fr::zero(); NUM_NEIGHBOURS];
+			for i in 0..NUM_NEIGHBOURS {
+				for j in 0..NUM_NEIGHBOURS {
+					new_s[i] += distributions[j][i];
+				}
+			}
+
+			s = new_s;
+		}
+
+		println!("new s: {:?}", s);
+
+		let mut sum = Fr::zero();
+		for x in s.iter() {
+			sum += x;
+		}
+		println!("sum: {:?}", sum);
+
+		s
+	}
+
+	fn filter_peers(
+		&self,
+	) -> (
+		[(PublicKey, Fr); NUM_NEIGHBOURS],
+		HashMap<PublicKey, Opinion>,
+	) {
+		let mut filtered_set: [(PublicKey, Fr); NUM_NEIGHBOURS] = self.set.clone();
+		let mut filtered_ops: HashMap<PublicKey, Opinion> = HashMap::new();
+
+		let mut valid_peers: Vec<PublicKey> = vec![];
+
+		// Convert the peer who didn't sign his opinion, to default peer
+		//
+		// Example:
+		//  In this set, peer3 didn't sign the opinion.
+		//		[(p1, 10), (p2, 10), (p3, 0)]
+		//   => [(p1, 10), (p2, 10), (null, 0)]
+		// Here, "null" means default peer
+		for i in 0..NUM_NEIGHBOURS {
+			let (pk, credits) = filtered_set[i].clone();
+			// If no credits, the peer did not sign his opinion
+			if credits == Fr::zero() {
+				filtered_set[i] = (PublicKey::default(), Fr::zero());
+			} else {
+				valid_peers.push(pk);
+			}
+		}
+
+		// Distribute the scores to valid peers
+		for i in 0..NUM_NEIGHBOURS {
+			let (pk_i, _) = filtered_set[i].clone();
+			if pk_i != PublicKey::default() {
+				let mut ops_i = self.ops.get(&pk_i).unwrap().clone();
+
+				// Give zero score for peers
+				//   - giving scores to itself
+				// 	 - giving scores to default peer
+				//
+				// Example:
+				//   Peer1 opinion
+				//		[(p1, 10), (p2, 10), (null, 10)]
+				// 	 => [(p1,  0), (p2, 10), (null,  0)]
+				//
+				for j in 0..NUM_NEIGHBOURS {
+					let (pk_j, score) = ops_i.scores[j].clone();
+					let true_pk = if i == j {
+						pk_i
+					} else if pk_j == pk_i {
+						PublicKey::default()
+					} else {
+						pk_j
+					};
+					let true_score = if true_pk == pk_i || true_pk == PublicKey::default() {
+						Fr::zero()
+					} else {
+						score
+					};
```

Comment:
Can we do 2 checks:
1) if pk_i == pk_j
2) if p_i == null

If any of these is true, replace the score with 0

Reply:
Yes, I think I over-complicated the logic here.
Please check update in https://github.com/eigen-trust/protocol/pull/154/commits/3e727c695a242bbd58c5b73102c3307a95f5a8eb